File: src/python/flows_specification/store.py
from __future__ import annotations
import hashlib
import json
import shutil
from pathlib import Path
from harness_engine import atomic_io

DIR = Path(".harness/specification/active")
PHASES = ("idea", "prd", "srs", "sdd", "review", "readiness", "approval")


def _path(name):
    return DIR / name


def canonical(value):
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"), sort_keys=True)


def digest(value):
    return "sha256:" + hashlib.sha256(canonical(value).encode()).hexdigest()


def load_run():
    try:
        return json.loads(_path("run.json").read_text())
    except Exception:
        return {
            "step": 0,
            "status": "in_progress",
            "phase": "start",
            "counters": {},
            "traceLabel": None,
            "terminalReason": None,
        }


def save_run(run):
    DIR.mkdir(parents=True, exist_ok=True)
    atomic_io.write_text_atomic(str(_path("run.json")), canonical(run))


def reset():
    if DIR.exists():
        shutil.rmtree(DIR)
    DIR.mkdir(parents=True, exist_ok=True)


def read_proposal(phase):
    try:
        return json.loads(_path(f"{phase}.proposal.json").read_text())
    except Exception:
        return None


def write_proposal(phase, value):
    DIR.mkdir(parents=True, exist_ok=True)
    atomic_io.write_text_atomic(str(_path(f"{phase}.proposal.json")), canonical(value))


def write_accepted(phase, value):
    DIR.mkdir(parents=True, exist_ok=True)
    atomic_io.write_text_atomic(str(_path(f"{phase}.accepted.json")), canonical(value))
    return digest(value)


def read_accepted(phase):
    try:
        value = json.loads(_path(f"{phase}.accepted.json").read_text())
        return value, digest(value)
    except Exception:
        return None, None


def bundle_digest():
    # Always joins exactly 5 segments in this fixed order (idea|prd|srs|sdd|readiness),
    # using "" for any phase not yet accepted — never omitting the position. An incomplete
    # chain still yields a stable (if unmatchable) digest instead of a shorter/misaligned
    # join. Mirrors SpecificationStore.BundleDigest() in the .NET port. Hashes the joined
    # string's raw UTF-8 bytes directly (not digest(), which JSON-encodes its argument
    # first) — same as SpecificationStore.Digest(string canonicalJson) in the .NET port.
    segments = [read_accepted(p)[1] or "" for p in ("idea", "prd", "srs", "sdd", "readiness")]
    joined = "|".join(segments)
    return "sha256:" + hashlib.sha256(joined.encode()).hexdigest()

File: src/rust/flows_specification/src/store.rs
//! Namespaced store (mirrors `SpecificationStore.cs`): persistence and digest helpers
//! shared by every phase of the Specification flow.

use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use std::{collections::HashSet, fs, path::Path};

pub(crate) const DIR: &str = ".harness/specification/active";
pub(crate) const SOURCES_FOLDER: &str = "specs/sources";

pub(crate) fn path(n: &str) -> String {
    format!("{DIR}/{n}")
}
pub(crate) fn read(n: &str) -> Option<Value> {
    serde_json::from_str(&fs::read_to_string(path(n)).ok()?).ok()
}
pub(crate) fn write(n: &str, v: &Value) {
    fs::create_dir_all(DIR).unwrap();
    harness_engine::atomic_io::write_atomic(
        Path::new(&path(n)),
        &serde_json::to_string(v).unwrap(),
    )
    .unwrap();
}
/// Digest of a JSON value's canonical (compact) serialization — used for parent-digest
/// freshness checks (ideaDigest, prdDigest, srsDigest, bundleDigest, ...).
pub(crate) fn digest(v: &Value) -> String {
    let mut h = Sha256::new();
    h.update(serde_json::to_vec(v).unwrap());
    format!("sha256:{:x}", h.finalize())
}
/// Digest of raw UTF-8 text — used for rendered Markdown content (never JSON-serialized),
/// mirroring `SpecificationPublisher.Digest(string content)`.
pub(crate) fn digest_str(content: &str) -> String {
    let mut h = Sha256::new();
    h.update(content.as_bytes());
    format!("sha256:{:x}", h.finalize())
}

pub(crate) fn run() -> Value {
    read("run.json")
        .unwrap_or(json!({"step":0,"status":"in_progress","phase":"start","counters":{}}))
}
/// Current harness step count (see `Harness.Engine.StateStore.Load().Step` in .NET) — synced
/// into run.json on every phase transition so `RunState.Step` reflects the real engine step,
/// not a locally-tracked counter.
fn current_step() -> i64 {
    harness_engine::state_store::load().step as i64
}
/// Normal phase transition: updates phase/status/step. Never touches `terminalReason` — a
/// non-terminal transition has nothing to report.
pub(crate) fn save(mut r: Value, p: &str, s: &str) {
    r["phase"] = json!(p);
    r["status"] = json!(s);
    r["step"] = json!(current_step());
    write("run.json", &r)
}
/// Terminal-status transition that does NOT change phase (mirrors the .NET `run with { Status
/// = ..., TerminalReason = ... }` pattern, which only overrides the fields listed). Records a
/// human-readable reason for `needs_human_decision`/`publish_blocked`.
pub(crate) fn save_status(mut r: Value, s: &str, reason: Option<&str>) {
    r["status"] = json!(s);
    r["step"] = json!(current_step());
    if let Some(reason) = reason {
        r["terminalReason"] = json!(reason);
    }
    write("run.json", &r)
}

pub(crate) fn proposal(p: &str) -> Option<Value> {
    read(&format!("{p}.proposal.json"))
}
pub(crate) fn accept(p: &str, v: &Value) {
    write(&format!("{p}.accepted.json"), v)
}
pub(crate) fn bundle_digest() -> String {
    let parts = ["idea", "prd", "srs", "sdd", "readiness"]
        .iter()
        .map(|phase| {
            read(&format!("{phase}.accepted.json")).map_or(String::new(), |value| digest(&value))
        })
        .collect::<Vec<_>>();
    let mut hasher = Sha256::new();
    hasher.update(parts.join("|"));
    format!("sha256:{:x}", hasher.finalize())
}

pub(crate) fn requirement_ids_of(srs: &Value) -> Vec<String> {
    srs["functionalRequirements"]
        .as_array()
        .into_iter()
        .flatten()
        .chain(srs["qualityRequirements"].as_array().into_iter().flatten())
        .filter_map(|item| item["id"].as_str().map(String::from))
        .collect()
}
pub(crate) fn adr_ids_of(sdd: &Value) -> Vec<String> {
    sdd["adrs"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|item| item["id"].as_str().map(String::from))
        .collect()
}

pub(crate) fn non_empty(value: &Value, key: &str) -> bool {
    value
        .get(key)
        .and_then(Value::as_array)
        .is_some_and(|items| !items.is_empty())
}
pub(crate) fn text_present(value: &Value, key: &str) -> bool {
    value
        .get(key)
        .and_then(Value::as_str)
        .is_some_and(|text| !text.trim().is_empty())
}
pub(crate) fn duplicate_ids(items: Option<&Vec<Value>>) -> bool {
    let mut ids = HashSet::new();
    items.is_some_and(|values| {
        values
            .iter()
            .filter_map(|item| item.get("id").and_then(Value::as_str))
            .any(|id| !ids.insert(id))
    })
}
